cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -320,4 +337,22 @@ private void popStash(@Nullable RevCommit stashID) throws GitAPIException {
     LOG.warn("Couldn't find Git stash with ID {} to drop. Ignoring missing stash.", stashID.getName());
   }
 
+  /**
+   * Does a <a href="https://git-scm.com/docs/git-reset"> hard reset </a> to the remote branch being tracked.
+   */
+  private String resetToCleanBranch(String remoteBranchName) throws IOException, GitAPIException {
+    List<RemoteConfig> remoteConfigs = git.remoteList().call();
+    if (remoteConfigs.size() != 1) {
+      throw new IllegalStateException(String.format("%s remotes found for git repository, expected 1.",
+                                                    remoteConfigs.size()));
+    }
+    RemoteConfig origin = remoteConfigs.get(0);
+    createCommand(git::fetch).setRefSpecs(origin.getFetchRefSpecs()).call();
+    git.reset()
+      .setMode(ResetCommand.ResetType.HARD)
+      .setRef(getRemoteBranchRefName(remoteBranchName, origin.getName()))
```

Comment:
We also have a TODO to fetch at this point. Is it going to be in a different PR?

Reply:
All the operations that are performed are on top top of the head of the default branch in the remote repo or the branch specified in the repository config stored for the CDAP namespace. Its not possible for someone to request a commit ID which isn't there in the repository manager unless either of the following are true:
1.  Commit is in the the future, i.e the remote head has moved forward and the caller got the commit ID from somewhere other than the repository manager.
2. Commit is in the history, i.e the remote history has changed (force push) so the commit ID is no longer a part of the remote branch that has been cloned.
3. Commit is in another branch.

To resolve 1, we can fetch and checkout the new head, but we have 2 problems:
*  Restoring the stashed changes can cause conflicts.
*  After going back to the previous commit once we have the file hash, we know that we're no longer at the head. The assumption of always being on the head is invalid. Future pushes will fail unless the caller does a cloneRemote() to get back to head and re-apply the changes.

To resolve 2, fetching the commit will help.
To resolve 3, the caller should create a new repository manager as the current implementation doesn't handle changing configs, eg: a push would result in pushing to the old branch.

Trying to handle 1 and 3 will make the implementation complex and will probably need the caller to do a re-clone / create a new manager later on anyway.
Situation 2 is unlikely, it would happen when the commit ID was stored somewhere and a re-clone was performed while the remote was force pushed. If the same repository manager provided the initial commit ID, it will still have the commit with it. If the commit ID is stored in persistent storage, callers can save the git hash along with it. 

Due to the additional edge cases, I think significantly its easier if the caller just re-clones and retries the entire pull/list apps flow again. Push will not face this problem since the new commit is made in the local repo.

Reply:
Well, it's the requirement of CDAP-20386 to handle (1). It's explicitly stated in the JIRA

Reply:
I also don't see what are the problems with doing a fetch and then checking out a proper commit in all cases. Given you know the commit you don't need to know if it's same branch, different branch or force pushed branch. You just reset the current local branch to the proper remote commit

Reply:
Could you elaborate on the problems? 

I don't think we plan to use stash in the local repo, do we? If so, what for?

Also you can't assume you are always on the remote branch head as there is always some time between the fetch and the push and remote branch can change in between. Local repo state does not matter here.

Reply:
And cloning won't fix it. I don't see how clone is different from fetch+checkout here

Reply:
The stashing and checking out of commits is removed. Earlier, to follow symlinks the method was checking out the commit. Now we aren't following symlinks. To restore the state before getting the file hash, the stashing was being done.